prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -802,22 +823,87 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
+                    }
+                );
+
+                if is_pure {
+                    let func_args = call_args.collect::<Vec<_>>();
+                    let pure_func_name = self.deps.require_ref::<crate::encoders::MirFunctionEncoder>(
+                        *func_def_id,
+                    ).unwrap().function_name;
+
+                    let pure_func_app = self.vcx.mk_func_app(pure_func_name, &func_args);
+
+                    let method_assign = {
+                        //TODO: Can we get `method_assign` in a better way? Maybe from the MirFunctionEncoder?
+                        let body = self.vcx.body.borrow_mut().get_impure_fn_body_identity(func_def_id.expect_local());
+                        let return_type = self.deps
+                            .require_ref::<crate::encoders::TypeEncoder>(body.return_ty())
+                            .unwrap();
+                        return_type.method_assign
+                    };
+
+                    self.stmt(vir::StmtData::MethodCall(self.vcx.alloc(vir::MethodCallData {
+                        targets: &[],
+                        method: method_assign,
+                        args: self.vcx.alloc_slice(&[dest, pure_func_app]),
+                    })));
+                }
+                else {
+                    let meth_args = std::iter::once(dest)
+                        .chain(call_args)
+                        .collect::<Vec<_>>();
+                    let func_out = self.deps.require_ref::<crate::encoders::MirImpureEncoder>(
+                        *func_def_id,
+                    ).unwrap();
+    
+                    self.stmt(vir::StmtData::MethodCall(self.vcx.alloc(vir::MethodCallData {
+                        targets: &[],
+                        method: func_out.method_name,
+                        args: self.vcx.alloc_slice(&meth_args),
+                    })));
+                }
 
-                let destination = self.encode_place(Place::from(*destination));
-                let args = args.iter().map(|op| self.encode_operand(op));
-                let args: Vec<_> = std::iter::once(destination).chain(args).collect();
-                self.stmt(vir::StmtData::MethodCall(self.vcx.alloc(vir::MethodCallData {
-                    targets: &[],
-                    method: func_out.method_name,
-                    args: self.vcx.alloc_slice(&args),
-                })));
                 self.vcx.alloc(vir::TerminatorStmtData::Goto(
                     self.vcx.alloc(vir::CfgBlockLabelData::BasicBlock(target.unwrap().as_usize())),
                 ))
             }
+            mir::TerminatorKind::Assert { cond, expected, msg, target, unwind } => {
+
+                let otherwise = match unwind {
+                    mir::UnwindAction::Cleanup(bb) => bb,
+                    _ => todo!()
+                };
+
+                let enc = self.encode_operand_snap(cond);
+                let enc = self.vcx.mk_func_app("s_Bool_val", &[enc]);
```

Comment:
I will be disabling the use of `mk_func_app` completely, you can replace a line like this with e.g. https://github.com/Aurel300/prusti-dev/blob/2ac6b644c7bb61763bcfb7f7671109e5eedaae49/prusti-encoder/src/encoders/mir_pure.rs#L549-L551 in the future